File: exercise/20/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.length() == 0) {
            return true;
        }
        stack<char> bracket;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                bracket.push(s[i]);
            } else if (!bracket.empty()) {
                if (s[i] == ')') {
                    if (bracket.top() == '(') {
                        bracket.pop();
                    } else {
                        return false;
                    }
                } else if (s[i] == ']') {
                    if (bracket.top() == '[') {
                        bracket.pop();
                    } else {
                        return false;
                    }
                } else if (s[i] == '}') {
                    if (bracket.top() == '{') {
                        bracket.pop();
                    } else {
                        return false;
                    }
                }
            } else {
                return false;
            }
        }
        return true;
    }
};

File: exercise/20/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 0:
            return False
        mapping = {'(': ')', '[': ']', '{': '}'}
        stack = []
        for c in s:
            if c in mapping:
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                if mapping[stack[-1]] == c:
                    stack.pop()
                else:
                    return False
        if len(stack) > 0:
            return False
        else:
            return True

